crabs/detection_tracking/inference_model.py

```diff
@@ -136,50 +153,173 @@ def prep_csv_writer(self):
-            (
-                frame_name,
-                frame.size,
-                '{{"clip":{}}}'.format("123"),
-                1,
-                0,
-                '{{"name":"rect","x":{},"y":{},"width":{},"height":{}}}'.format(
-                    xmin, ymin, width_box, height_box
-                ),
-                '{{"track":{}}}'.format(id),
+        mota_values = []
+        prev_frame: Union[Optional[List[List[float]]], None] = None
+        for gt_boxes, tracked_boxes in zip(gt_boxes_list, tracked_boxes_list):
+            mota = evaluate_mota(
+                gt_boxes, tracked_boxes, iou_threshold, prev_frame
             )
-        )
+            mota_values.append(mota)
+            # Update previous frame IDs for the next iteration
+            prev_frame = [list(box) for box in tracked_boxes]
+
+        return mota_values
+
+    def get_prediction(self, frame: np.ndarray) -> torch.Tensor:
+        """
+        Get prediction from the trained model for a given frame.
+
+        Parameters:
+        -----------
+        frame : np.ndarray
+            The input frame for which prediction is to be obtained.
+
+        Returns:
+        --------
+        torch.Tensor:
+            The prediction tensor from the trained model.
+        """
+        transform = transforms.Compose([transforms.ToTensor()])
+        img = transform(frame).to(self.args.accelerator)
+        img = img.unsqueeze(0)
+        return self.trained_model(img)
+
+    def update_tracking(self, prediction: dict) -> List[List[float]]:
+        """
+        Update the tracking system with the latest prediction.
+
+        Parameters:
+        -----------
+        prediction : dict
+            Dictionary containing predicted bounding boxes, scores, and labels.
+
+        Returns:
+        --------
+        List[List[float]]:
+            List of tracked bounding boxes after updating the tracking system.
+        """
+        pred_sort = self.prep_sort(prediction)
+        tracked_boxes = self.sort_tracker.update(pred_sort)
+        self.tracked_list.append(tracked_boxes)
+        return tracked_boxes
+
+    def handle_output(
+        self,
+        tracked_boxes: List[List[float]],
+        frame: np.ndarray,
+        frame_number: int,
+    ) -> None:
+        """
+        Handle the output based argument options.
+
+        Parameters:
+        -----------
+        tracked_boxes : List[List[float]]
+            List of tracked bounding boxes.
+        frame : np.ndarray
+            The current frame.
+        frame_number : int
+            The frame number.
+        """
+        if self.args.save_csv_and_frames:
+            if self.args.save_video:
+                frame_copy = save_frame_and_csv(
+                    self.video_file_root,
+                    self.tracking_output_dir,
+                    tracked_boxes,
+                    frame,
+                    frame_number,
+                    self.csv_writer,
+                )
```

Comment:
I don't understand a few things here:
- Why are the if cases for `self.args.save_csv_and_frames:` and `self.args.save_video` nested? The don't seem to do related things right?
- I understand that if you call `save_frame_and_csv` with the default value for `save_plot`, the bounding boxes are plotted in the saved frame...so why do you need to call `draw_bbox` afterwards?

I may be missing something but I think it could be subs with this?
```
if self.args.save_csv_and_frames:
	save_frame_and_csv(
	    self.video_file_root,
	    self.tracking_output_dir,
	    tracked_boxes,
	    frame,
	    frame_number,
	    self.csv_writer,
	    save_plot=False,
	)

if self.args.save_video:
	frame_copy = save_frame_and_csv(
	    self.video_file_root,
	    self.tracking_output_dir,
	    tracked_boxes,
	    frame,
	    frame_number,
	    self.csv_writer,
	    save_plot=True,
	)
	self.out.write(frame_copy)
```

Reply:
The nested coming from our previous pair programming and I thought it might be useful. The save plot here is saving the tracked video with bounding box. not saving frame images with bounding box. (Though you were right the `draw_bbox` shouldn't be there)
When we run this in the future, we might want to have both, save the track in csv for correcting them, and save the tracked video for visual inspection etc.
Hence the second option is there for only saving the csv. 
Happy to change it to two separate if statement instead.

Reply:
changed now, much cleaner now